examples/ensemble/plot_gradient_boosting_early_stopping.py

```diff
@@ -39,131 +43,135 @@
+# to assess the performance and convergence of the models.
+
+train_errors_without = []
+val_errors_without = []
+
+train_errors_with = []
+val_errors_with = []
+
+for i, (train_pred, val_pred) in enumerate(
+    zip(
+        gbm_no_early_stopping.staged_predict(X_train),
+        gbm_no_early_stopping.staged_predict(X_val),
+    )
+):
+    train_errors_without.append(mean_squared_error(y_train, train_pred))
+    val_errors_without.append(mean_squared_error(y_val, val_pred))
+
+for i, (train_pred, val_pred) in enumerate(
+    zip(
+        gbm_with_early_stopping.staged_predict(X_train),
+        gbm_with_early_stopping.staged_predict(X_val),
+    )
+):
+    train_errors_with.append(mean_squared_error(y_train, train_pred))
+    val_errors_with.append(mean_squared_error(y_val, val_pred))
 
-plt.xlabel("Datasets")
-plt.ylabel("Test score")
+# %%
+# Visualize Comparision
+# ---------------------
+# It includes three subplots:
+# 1. Plotting training errors of both models over boosting iterations.
+# 2. Plotting validation errors of both models over boosting iterations.
+# 3. Creating a bar chart to compare the training times and the estimator used
+# of the models with and without early stopping.
+
+plt.figure(figsize=(15, 6))
+
+plt.subplot(1, 3, 1)
+plt.plot(train_errors_without, label="Training Error (No Early Stopping)")
+plt.plot(train_errors_with, label="Training Error (With Early Stopping)")
+plt.xlabel("Boosting Iterations")
+plt.ylabel("MSE (Training)")
+plt.legend()
+
+plt.subplot(1, 3, 2)
+plt.plot(val_errors_without, label="Validation Error (No Early Stopping)")
+plt.plot(val_errors_with, label="Validation Error (With Early Stopping)")
+plt.xlabel("Boosting Iterations")
+plt.ylabel("MSE (Validation)")
+plt.legend()
+
+plt.subplot(1, 3, 3)
+training_times = [training_time_without, training_time_with]
+labels = ["No Early Stopping", "With Early Stopping"]
+bars = plt.bar(labels, training_times)
+plt.ylabel("Training Time (s)")
+
+for bar, n_estimators in zip(bars, [estimators_without, estimators_with]):
+    height = bar.get_height()
+    plt.text(
+        bar.get_x() + bar.get_width() / 2,
+        height + 0.01,
+        f"Estimators: {n_estimators}",
+        ha="center",
+        va="bottom",
+    )
 
+plt.tight_layout()
 plt.show()
 
-
 # %%
-# Compare fit times with and without early stopping
-# -------------------------------------------------
-
-plt.figure(figsize=(9, 5))
-
-bar1 = plt.bar(
-    index, time_gb, bar_width, label="Without early stopping", color="crimson"
-)
-bar2 = plt.bar(
-    index + bar_width, time_gbes, bar_width, label="With early stopping", color="coral"
-)
-
-max_y = np.amax(np.maximum(time_gb, time_gbes))
-
-plt.xticks(index + bar_width, names)
-plt.yticks(np.linspace(0, 1.3 * max_y, 13))
-
-autolabel(bar1, n_gb)
-autolabel(bar2, n_gbes)
-
-plt.ylim([0, 1.3 * max_y])
-plt.legend(loc="best")
-plt.grid(True)
-
-plt.xlabel("Datasets")
-plt.ylabel("Fit Time")
-
-plt.show()
+# Summary
```

Comment:
Before the summary, add a comment below the figure on why the training error is not the same for both models.